File: pattern_java/18memento/5/Originator.java


///融合备忘录的发起人角色
//增加了clone方法，产生了一个备份对象，需要使用的时候再还原
public class Originator implements Cloneable{
    
    //内部状态
    private String state = "";
    
    public String getState() {
        return state;
    }

    public void setState(String state) {
        this.state = state;
    }

    //创建一个备忘录
    public Originator createMemento(){
        return this.clone();
    }
    
    //恢复一个备忘录
    public void restoreMemento(Originator _originator){
        this.setState(_originator.getState());
    }
    
    //克隆当前对象
    @Override
    protected Originator clone(){
        
        try {
            return (Originator)super.clone();
        } catch (CloneNotSupportedException e) {
            e.printStackTrace();
        }
        return null;
    }
}

File: pattern/laravel/src/25Memento/Originator.php
    <?php
    
    namespace Memento;
    
    class Originator
    {
        /* @var mixed */
        private $state;
    
        // 这个类还可以包含不属于备忘录状态的额外数据
    
        /**
         * @param mixed $state
         */
        public function setState($state)
        {
            // 必须检查该类子类内部的状态类型或者使用依赖注入
            $this->state = $state;
        }
    
        /**
         * @return Memento
         */
        public function getStateAsMemento()
        {
            // 在Memento中必须保存一份隔离的备份
            $state = is_object($this->state) ? clone $this->state : $this->state;
    
            return new Memento($state);
        }
    
        public function restoreFromMemento(Memento $memento)
        {
            $this->state = $memento->getState();
        }
    }